sgkit/utils.py

```diff
@@ -194,3 +195,32 @@ def split_array_chunks(n: int, blocks: int) -> Tuple[int, ...]:
     n_div, n_mod = np.divmod(n, blocks)
     chunks = n_mod * (n_div + 1,) + (blocks - n_mod) * (n_div,)
     return chunks  # type: ignore[no-any-return]
+
+
+def max_str_len(a: ArrayLike) -> da.Array:
+    """Compute maximum string length for elements of an array
+
+    Parameters
+    ----------
+    a
+        Array of any shape, must have string or object dtype
+
+    Returns
+    -------
+    max_length
+        Dask array with scalar integer value
+    """
+    if a.size == 0:
+        raise ValueError("Max string length cannot be calculated for empty array")
+    if a.dtype.kind == "O":
+        a = a.astype(str)
+    if a.dtype.kind not in {"U", "S"}:
+        raise ValueError(f"Array must have string dtype (got dtype {a.dtype})")
+
+    ndim = a.ndim
+
+    def fn(x: np.ndarray) -> np.ndarray:
+        max_len = np.asarray(np.frompyfunc(len, 1, 1)(x)).max()
+        return np.expand_dims(max_len, list(range(ndim)))
+
+    return da.map_blocks(fn, a, chunks=(1,) * ndim, dtype=int).max()
```

Comment:
is this significantly faster than:

```
return da.frompyfunc(len, 1, 1)(a).max()
```

Reply:
Doesn't look like it could be based on https://github.com/dask/dask/blob/7a46e7b4a436f5152872e6d7fa4f5291342bdd2f/dask/array/ufunc.py#L47 since it's wrapping np.frompyfunc.  It's about the same, maybe slightly slower in this test:

```python
def max_str_len1(a):
    ndim = a.ndim

    def fn(x: np.ndarray) -> np.ndarray:
        max_len = np.asarray(np.frompyfunc(len, 1, 1)(x)).max()
        return np.expand_dims(max_len, list(range(ndim)))

    return da.map_blocks(fn, a, chunks=(1,) * ndim, dtype=int).max()

def max_str_len2(a):
    ndim = a.ndim
    str_len = da.frompyfunc(len, 1, 1)
    
    def fn(x: np.ndarray) -> np.ndarray:
        max_len = np.asarray(str_len(x)).max()
        return np.expand_dims(max_len, list(range(ndim)))

    return da.map_blocks(fn, a, chunks=(1,) * ndim, dtype=int).max()

x = da.asarray(['x']*100000, chunks=(100))
%timeit max_str_len1(x).compute()
154 ms ± 1.23 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit max_str_len2(x).compute()
155 ms ± 2.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Apparently not defining it per-block doesn't matter either.  I started a list in https://github.com/pystatgen/sgkit/issues/68 though with this function on it as a fast one that could be easily tracked.

Reply:
No, no, I mean the whole implementation would be:

```
def max_str_len(a: ArrayLike) -> ArrayLike:
    if a.size == 0:
        raise ValueError("Max string length cannot be calculated for empty array")
    if a.dtype.kind == "O":
        a = a.astype(str)
    if a.dtype.kind not in {"U", "S"}:
        raise ValueError(f"Array must have string dtype (got dtype {a.dtype})")

    lens = np.frompyfunc(len, 1, 1)(a)
    return lens if isinstance(lens, int) else lens.max()
```

Notice that:
 * we return `ArrayLike` instead of forcing Dask Array
 * we use `np` dispatching
 * we need to handle numpy scalars (e.g. `np.asarray("foo")`)
 * for the implementation above you need to adjust tests to call `compute()` only on Dask Arrays, or call Dask aware equality check

Reply:
Ahh I see what you mean -- yea dask interprets them as more or less the same graph and there's essentially no difference in times between them that I can find.  I don't like the numpy semantics since dask, xarray and cupy preserve the array backend for ufunc_reduce functions on scalars, so I made that a special case: 
 https://github.com/pystatgen/sgkit/pull/314/commits/ed9c0703d9f46e4c8afd938901b430495a8f0b7a#diff-ee100e898f8291cbcc9fcecc41cef338R219

This means the function preserves the array type and can be forced to evaluate (for dask or xarray + dask) with just `int(max_str_len(x))` rather than needing to switch on anything in the calling code.